### PR TITLE
ENH: add a stack of reversal functions to linprog

### PR DESCRIPTION
**Pull request by janvle**
#### Reference issue
Second step to implement suggestion #11570

#### What does this implement/fix?
This PR does the groundwork for a modular concept for _presolve().

#### Additional information
This is a set of interrelated changes. I'll give short descriptions below. At a couple of locations, testing is possible; I have indicated these locations as well.

(1) Repair aliasing in `bounds` in `_presolve()`
This goes unnoticed because in `_linprog()`, the call to `_presolve()` uses the same `lp` as input as well as output. But still.

(2) Eliminate the `_postprocess()` function
The function is only used once, in `_linprog()`, and consists of three calls, to `_postsolve()`, to `_check_result()` and to `_display_summary()`. `_postsolve()` returns `bounds` which is input to `_check_result()`.
In the new setup `bounds` will be reduced along with the other problem parameters. `_check_result()` needs the original bounds, but it is a waste of effort to reconstruct the original `bounds` in `_postsolve()`, because the original value is available (`lp_o` in `_linprog()`). So I replaced the `_postprocess()` call in '_linprog()` by its three constituents, which then allows to remove `bounds` as output of `_postsolve()` and use the stored original `bounds` as input to `_check_result()`.

(3) Reduce the `bounds` variable along with and matching changes in the other problem parameters
Return modified `bounds` from `_presolve()`. In `_get_Abc()`, remove the reduction of `bounds` and add a `.copy()`, since `bounds` gets changed.

If I reconstruct the original `bounds` on `_postsolve()`, so that a the code can do what it did and return the original `bounds`, these changes passes `test_linprog.py`

(4) Eliminate the `undo` argument from `_get_Abc()`
The `undo` argument is superfluous now that `_get_Abc()` gets the reduced `bounds` as input.

Question: does the comment in _get_Abc() line 1096
    `# if preprocessing is on, lb == ub can't happen`
    `# if preprocessing is off, then it would be best to convert that`
    `# to an equality constraint, but it's tricky to make the other`
    `# required modifications from inside here.`
still make sense? 

(5) Restructure the procedure by which `_postsolve()` reverses the effects of `_presolve()` and `_get_Abc()`. Use the opposite order: so first reverse what `_get_Abc()` did, then what `_presolve()` did.

(6) Stop returning `bounds` from `_postsolve()`
Reconstruction is superfluous, see (2). Update the locations which call `_postsolve()` in the solvers and in `_linprog()`.
Update the call to `_check_result()`: input the original bounds.

Question: can `_postsolve()` input argument `copy` be removed? Why is it there? Only used (set to True) in `_linprog_rs.py`, `_phase_two()`.

A commit after these changes passes `test_linprog.py`.

(7) Reinsert the bounds test which was removed from `_clean_inputs()`
Add the check for infeasible bounds (ub < lb, ub = -inf, lb = inf) to `_presolve()`.
Add a test function for this to `test_linprog.py`: `test_bounds_infeasible`.

Question: moving the check from `_clean_inputs()` to `_presolve()` implies that with the presolve option set to `False`, no infeasibility tests are carried out (contrary to earlier). That is logical, but infeasibility is not always detected by the solver (see test function, simplex method). Any thouhgts on this?

A commit after these changes passes `test_linprog.py`.

(8) Change the `undo` mechanism.
The variable `undo` is now only present and used in `_presolve()` and in `_postsolve()`.
Change `undo` into a list of reversal functions (for now: just a single one), which reinsert into `x` what was removed by `_presolve()`.

A commit after these changes passes `test_linprog.py`.
Benchmarks show no change in performance.

### CODE REVIEW

**Comment by mdhaber** [comment]
@janvle I don't think the failures are your fault. I think tests have been running slowly on Azure in general for some reason. I'll take a closer look at your changes soon.

File: scipy/optimize/_linprog.py
Comment (mdhaber) [inline scipy/optimize/_linprog.py]: Just checking understanding - there was no reason to reconstruct the bounds from `undo` because the unmodified bounds are just available in `lp_o`, and there was no other reason to perform the reconstruction?

File: scipy/optimize/_linprog.py
Comment (mdhaber) [inline scipy/optimize/_linprog.py]: `tol` is no longer an argument of `_postsolve`, though. Looks like it's going in as `copy`. But it looks like `copy` is not used anymore, either.

File: scipy/optimize/_linprog_ip.py
Comment (mdhaber) [inline scipy/optimize/_linprog_ip.py]: I see that `tol` was not used at all in `_postsolve`

File: scipy/optimize/_linprog_rs.py
Comment (mdhaber) [inline scipy/optimize/_linprog_rs.py]: Looks like `copy` can be removed, too.

File: scipy/optimize/_linprog_util.py
Comment (mdhaber) [inline scipy/optimize/_linprog_util.py]: They don't anymore though, right? You created a copy.

File: scipy/optimize/_linprog_util.py
Comment (mdhaber) [inline scipy/optimize/_linprog_util.py]: I'm forgetting - is this being moved from somewhere else? But I don't see it deleted anywhere else in the changes.

Also, shouldn't the comparison be string inequality? Upper and lower bounds equal is just an equality constraint. The variable is fixed and can be removed from the problem (a presolve operation).

**Review by mdhaber — CHANGES_REQUESTED** [review]
I'm going to have to pause the review there for now. Looking good! But answers to the questions so far will help me finish.

File: scipy/optimize/_linprog.py
Comment (janvle) [reply scipy/optimize/_linprog.py:511]: Of course.

File: scipy/optimize/_linprog.py
Comment (janvle) [reply scipy/optimize/_linprog.py:578]: The `tol` argument should be removed here.
I don;t know about `copy`: it is only used once (set to `True`) in `_linprog_rs.py`, `_phase_two()`. I haven't dug into what this exactly does. Do you know?

File: scipy/optimize/_linprog_ip.py
Comment (janvle) [reply scipy/optimize/_linprog_ip.py:711]: Yes, the `tol`-argument is not used in `_postsolve()`.

File: scipy/optimize/_linprog_rs.py
Comment (janvle) [reply scipy/optimize/_linprog_rs.py:347]: It certainly looks that way. Do you remember why it was in there?

File: scipy/optimize/_linprog_util.py
Comment (janvle) [reply scipy/optimize/_linprog_util.py:658]: Yes. I'll remove the comments.

File: scipy/optimize/_linprog_util.py
Comment (janvle) [reply scipy/optimize/_linprog_util.py:590]: This check was part of `_clean_inputs()`, it was removed there since I'd like to have all infeasibility tests in one location. And you are right: there must not be an equal sign.

Comment (janvle) [reply scipy/optimize/_linprog_util.py:590]: On second thought: why did this pass the tests? I'll take a closer look at this.

File: scipy/optimize/_linprog.py
Comment (mdhaber) [reply scipy/optimize/_linprog.py:578]: I don't remember. I'll take another look when you'd like me to review again.

File: scipy/optimize/_linprog_util.py
Comment (janvle) [reply scipy/optimize/_linprog_util.py:590]: There were no tests which explicitly checked fixed bounds. I have added one.

File: scipy/optimize/_linprog.py
Comment (janvle) [reply scipy/optimize/_linprog.py:578]: I'll remove the parameter completely.
I went through your comments and made the necessary changes. Do you prefer me to push the results before you continue reviewing?

File: scipy/optimize/_linprog_rs.py
Comment (mdhaber) [reply scipy/optimize/_linprog_rs.py:347]: Yes, now I do. When the option `disp=True` or when using the callback, `_postsolve` gets called every iteration. When originally written, `_postsolve` was only supposed to be run "post-solve", not every iteration, so it added variables back into the solution vector `x` in-place. Interior point didn't care, because it actually passes in `x/tau`, and legacy simplex effectively copies with some fancy indexing before passing `x` in, but `revised simplex` didn't copy, so we (I?) added the argument, and `_postsolve` performed the copy only if necessary - only if variables would be added back. But now I think `_unscale` (which was added later) guarantees that a copy is made, so I think it's safe to remove.

File: scipy/optimize/_linprog.py
Comment (mdhaber) [reply scipy/optimize/_linprog.py:578]: Please go ahead and push and I'll review.

File: scipy/optimize/tests/test_linprog.py
Comment (mdhaber) [inline scipy/optimize/tests/test_linprog.py:267]:
```suggestion
```

File: scipy/optimize/tests/test_linprog.py
Comment (mdhaber) [inline scipy/optimize/tests/test_linprog.py:394]: We're discussing this in gh-11618; I hope it is going to be fixed.

File: scipy/optimize/tests/test_linprog.py
Comment (mdhaber) [inline scipy/optimize/tests/test_linprog.py:365]: It looks like you might have put this in a function to avoid the lines getting too long?
Consider [parameterizing the test](https://docs.pytest.org/en/latest/parametrize.html). I think it will accomplish the same thing and make it more concise. Maybe also check the number of iterations to ensure that this is solved in presolve?

File: scipy/optimize/tests/test_linprog.py
Comment (mdhaber) [inline scipy/optimize/tests/test_linprog.py:375]: Same as above; consider parameterizing and checking that the problem is solved in presolve.

File: benchmarks/asv.conf.json
Comment (mdhaber) [inline benchmarks/asv.conf.json:32]: Please revert changes to benchmark files.

File: scipy/optimize/_linprog_util.py
Comment (mdhaber) [inline scipy/optimize/_linprog_util.py:590]: Thanks for remembering to put this back : ) I agree it belongs here.

File: scipy/optimize/_linprog_util.py
Comment (mdhaber) [inline scipy/optimize/_linprog_util.py:1072]: Could you explain this?

File: scipy/optimize/_linprog_util.py
Comment (mdhaber) [inline scipy/optimize/_linprog_util.py:1070]: Curious why these are removed?

File: scipy/optimize/_linprog_util.py
Comment (mdhaber) [inline scipy/optimize/_linprog_util.py:763]: Why not `np.arange(N)` so you can just do `i - index_offset` below?

File: scipy/optimize/_linprog_util.py
Comment (mdhaber) [inline scipy/optimize/_linprog_util.py:766]: Why did you need to use `astype(float)`?

File: scipy/optimize/_linprog_util.py
Comment (mdhaber) [inline scipy/optimize/_linprog_util.py:767]:
```suggestion
            return x_rev
```

File: scipy/optimize/_linprog_util.py
Comment (mdhaber) [inline scipy/optimize/_linprog_util.py:755]: Clearly I didn't know about `np.insert` when I wrote the original for loop that re-inserted variables. Also, I like `np.flatnonzero`. I've always used `np.where` but then I'm only ever interested in the 0th element of the result. This is much cleaner.

**Review by mdhaber — CHANGES_REQUESTED** [review]
Looking pretty good! A few comments.

**Comment by mdhaber** [comment]
> Question: moving the check from _clean_inputs() to _presolve() implies that with the presolve option set to False, no infeasibility tests are carried out (contrary to earlier). That is logical, but infeasibility is not always detected by the solver (see test function, simplex method). Any thouhgts on this?

If it's only `simplex` that has has trouble without `presolve`, I'm not concerned. We don't really need to maintain that anymore.

**Comment by mdhaber** [comment]
> Question: does the comment in _get_Abc() line 1096 still make sense

Ooops, I asked you why you removed it. I think without presolve, equal upper and lower bounds are treated as two separate inequality constraints in `_get_Abc`. I was probably suggesting that they be turned into one equality constraint rather than two constraints with slack variables. I suppose I thought this was reasonable to do in `_get_Abc` even if presolve is off, and perhaps I wanted to come back and implement that. But I don't think it's terribly important.

File: scipy/optimize/_linprog.py
Comment (janvle) [reply scipy/optimize/_linprog.py]: Sure!

File: scipy/optimize/_linprog.py
Comment (janvle) [reply scipy/optimize/_linprog.py]: That is right.

File: scipy/optimize/_linprog_rs.py
Comment (janvle) [reply scipy/optimize/_linprog_rs.py]: I removed it.

File: scipy/optimize/tests/test_linprog.py
Comment (janvle) [reply scipy/optimize/tests/test_linprog.py:267]: OK

File: scipy/optimize/tests/test_linprog.py
Comment (janvle) [reply scipy/optimize/tests/test_linprog.py:365]: Actually I just copied the style of the neighboring tests. But I see that these function definitions are only used to check is errors or warnings are raised. I'll stick to that.

Comment (janvle) [reply scipy/optimize/tests/test_linprog.py:365]: And I'll include a check on `res.nit == 0`.

File: scipy/optimize/tests/test_linprog.py
Comment (janvle) [reply scipy/optimize/tests/test_linprog.py:394]: That would simplify this test. I'll refer to this issue in the test.

File: scipy/optimize/tests/test_linprog.py
Comment (janvle) [reply scipy/optimize/tests/test_linprog.py:375]: See above. Changed it so that the function is only used where errors are expected.

File: benchmarks/asv.conf.json
Comment (janvle) [reply benchmarks/asv.conf.json:32]: I did not notice the file was in the commit. My benchmarks need Cython 0.29.14. I'll revert & keep my change local.

File: scipy/optimize/_linprog_util.py
Comment (janvle) [reply scipy/optimize/_linprog_util.py:1072]: I'd better write a clearer comment :-)
What I mean to say is that `lbs` and `ubs`, introduced 2 lines down, may change, and these changes feed back into `bounds`.

File: scipy/optimize/_linprog_util.py
Comment (janvle) [reply scipy/optimize/_linprog_util.py:1070]: Mainly because I think `_get_Abc()` should just convert the problem to standard form. I can see that it is tempting to turn `lb == ub` into an equality constraint here, but that would slip presolve functionality into `_get_Abc()`. I think if one would want to prevent spurious inequalities, one would just enable presolve.

File: scipy/optimize/_linprog_util.py
Comment (janvle) [reply scipy/optimize/_linprog_util.py:763]: Yes, why not, it is much cleaner.

File: scipy/optimize/_linprog_util.py
Comment (janvle) [reply scipy/optimize/_linprog_util.py:766]: If `x-mod` is an integer array, inserting floats does not turn it into a float array. Instead, inserted floats are rounded; the type of x_mod is leading. I have looked for better locations to prevent this, but did not find any logical location.

File: scipy/optimize/_linprog_util.py
Comment (janvle) [reply scipy/optimize/_linprog_util.py:767]: OK

**Comment by janvle** [comment]
OK, let's proceed. I'm ready to PR a `_presolve()` implementation with subfunctions.

**Comment by mdhaber** [comment]
> OK, let's proceed.

? I don't see any changes. Did you mean to push? At least the changes to benchmarks need to be reverted. (And yes, I have to do the same thing. I've suggested bumping those versions, but someone disagreed.)

**Comment by janvle** [comment]
No, I meant to proceed with _this_ PR. Changes as a result of your review do not need a new push, so I wait for your further remarks.
The PR I was referring to is the next phase. It is ready to go once this phase is complete.

**Comment by mdhaber** [comment]
> Changes as a result of your review do not need a new push

I'm sorry, but I don't understand. I've reviewed this twice, once [at the end of March](https://github.com/scipy/scipy/pull/11691#issuecomment-606499880) and [once in mid-April](https://github.com/scipy/scipy/pull/11691#pullrequestreview-391821937). I paused the first review asking for clarification before I could continue. You responded, and I continued and finished my review in the second review. I didn't intend to make any other comments. Once you push, if your fixes look good and I don't find anything new, I'll merge. Does that make sense?

**Comment by janvle** [comment]
Yes, I'll push the last changes. I didn't realize your review was complete.

**Comment by janvle** [comment]
@mdhaber: Some of the automatic checks fail, but I cannot make sense of the logs. Can you?

**Comment by mdhaber** [comment]
It's probably not your fault. There have been a lot of CI issues lately. Not at a computer now so I can't check the status.

**Comment by tylerjereddy** [comment]
Correct, Azure CI Windows fails are not related to this PR

**Review by mdhaber — COMMENTED** [review]
I'm going to close and reopen to retrigger CI. I want to see code coverage in Azure, but it's not available because the builds failed.

File: scipy/optimize/tests/test_linprog.py
Comment (mdhaber) [reply scipy/optimize/tests/test_linprog.py:385]: OK I was also suggesting you parameterize these to make the test less repetitive:
https://docs.pytest.org/en/latest/parametrize.html
`c`, `bounds`, the optimal objective, and the solution would be parameters.

Comment (mdhaber) [reply scipy/optimize/tests/test_linprog.py:385]: If you address this and the (few) other new comments and if code coverage looks good, I think I can merge.

File: scipy/optimize/tests/test_linprog.py
Comment (mdhaber) [inline scipy/optimize/tests/test_linprog.py:360]: I'd like to see a more involved test that has constraints, because none of these actually run the main linprog methods due to:
```
    # no constraints indicates that problem is trivial
    if A_eq.size == 0 and A_ub.size == 0:
         # code that solves trivial problems
```

**Comment by mdhaber** [comment]
@janvle the test failure is just a timeout, not your fault.

**Comment by janvle** [comment]
I have been looking into ways to parametrize the tests, as you suggested. Took me some time to get into the details. What I think we need is a way so set `linprog()` parameters at the beginning of a test, without having to repeat them if the test calls `linprog()` repeatedly. The [existing pytest-option](https://docs.pytest.org/en/latest/parametrize.html) does not seem to do this, at least I do not get it to work and the added complexity is a disadvantage if it would work.

This may work:

- add the `linprog()` parameters as properties to `LinprogCommonTests`
- define three functions in `LinprogCommonTests`: one to reset these properties, one to change specific properties, and one to call `linprog()` with these properties as arguments. Not involved, see code below.

I see that at line 225, global arguments are defined for a couple of `linprog()` arguments; these work in a more or less similar fashion.
With these functions, each test starts by setting test-wide arguments, and then for each call to `linprog()`, you only set those arguments which change.

What do you think?
Would it be worthwhile to adapt all the tests in this way?



    def run_linprog(self):
        # call linprog() with preset parameters
        return linprog(c=self.c, A_ub=self.A_ub, b_ub=self.b_ub, A_eq=self.A_eq,
                       b_eq=self.b_eq, bounds=self.bounds, method=self.method,
                       callback=self.callback, options=self.options, x0=self.x0)

    def set_linprog(self, c=None, A_ub=None, b_ub=None, A_eq=None,
                    b_eq=None, bounds=None, method=None,
                    callback=None, options=None, x0=None):
        # set linprog() input parameters
        # keep current values of parameters with None value
        if c is not None:
            self.c = c
        if A_ub is not None:
            self.A_ub = A_ub
        if b_ub is not None:
            self.b_ub = b_ub
        if A_eq is not None:
            self.A_eq = A_eq
        if b_eq is not None:
            self.b_eq = b_eq
        if bounds is not None:
            self.bounds = bounds
        if method is not None:
            self.method = method
        if callback is not None:
            self.callback = callback
        if options is not None:
            self.options = options
        if x0 is not None:
            self.x0 = x0

    def reset_linprog(self, c=None, A_ub=None, b_ub=None, A_eq=None,
                      b_eq=None, bounds=None, method=None,
                      callback=None, options=None, x0=None):
        # reset linprog() input parameters
        # initialize all parameters to either given value or None
        self.c = c
        self.A_ub = A_ub
        self.b_ub = b_ub
        self.A_eq = A_eq
        self.b_eq = b_eq
        self.bounds = bounds
        self.method = method
        self.callback = callback
        self.options = options
        self.x0 = x0

File: scipy/optimize/tests/test_linprog.py
Comment (janvle) [reply scipy/optimize/tests/test_linprog.py:360]: No problem.

**Comment by mdhaber** [comment]
Hi @janvle, I'm sorry, but my reviews are backlogged due to a very busy quarter. I don't want to add that sort of change to all the tests with this PR, but if you need to add something like that just for the new tests, that is OK.

**Comment by janvle** [comment]
OK, I'll leave the tests alone, just adding a more involved test for the fixed bounds case. I pushed & now I'm taking a short vacation :-).
If all is well, this PR can be merged, and then I'll prepare a new PR for the presolve functionality.

**Comment by mdhaber** [comment]
Ok, if you're away, I'll remove the benchmarks file and address anything else that remains before merging.

**Comment by janvle** [comment]
@mdhaber Did you succeed? I'm ready to create a PR for the last phase.

**Comment by mdhaber** [comment]
Code coverage is good. Running benchmarks.

**Comment by mdhaber** [comment]
Thanks @janvle - and thanks for your patience.

**Pull request merged**
